proxy/http/http_test.go

```diff
@@ -0,0 +1,375 @@
+	}
+	if bw.Buffered() == 0 {
+		t.Fatalf("Cant't write header to bufio writer")
+	}
+}
+
+func TestHTTPRequestError(t *testing.T) {
+	errorReq := "/ HTTP/1.1\r\n" +
+		"Host: localhost:10000\r\n" +
+		"\r\n"
+	rightReq := "GET / HTTP/1.1\r\n" +
+		"Host: localhost:10000\r\n" +
+		"\r\n"
+	req := &Request{}
+	br := bufio.NewReader(strings.NewReader(errorReq))
+	err := req.ReadFrom(br)
+	if err == nil {
+		t.Fatal("expected error: fail to read start line of request")
+	}
+	if !strings.Contains(err.Error(), "fail to read start line of request") {
+		t.Fatalf("unexpected error: %s", err.Error())
+	}
+
+	req.Reset()
+	nbr := bufio.NewReader(strings.NewReader(rightReq))
+	req.ReadFrom(nbr)
+	err = req.ReadFrom(nbr)
+	if err == nil {
+		t.Fatal("expected error: request already initialized")
+	}
+	if !strings.Contains(err.Error(), "request already initialized") {
+		t.Fatalf("unexpected error: %s", err.Error())
+	}
+
+	req.Reset()
+	err = req.ReadFrom(nil)
+	if err == nil {
+		t.Fatal("nil reader provided")
+	}
+	if !strings.Contains(err.Error(), "nil reader provided") {
+		t.Fatalf("unexpected error: %s", err.Error())
+	}
+	w := bytebufferpool.Get()
+	bw := bufio.NewWriter(w)
+	err = req.WriteHeaderTo(bw)
+	if err == nil {
+		t.Fatal("expected error:Empty request, nothing to write")
+	}
+	if !strings.Contains(err.Error(), "Empty request, nothing to write") {
+		t.Fatalf("unexpected error: %s", err)
+	}
+}
+
+func TestHTTPResponse(t *testing.T) {
+	s := "HTTP/1.1 200 ok\r\n" +
+		"Cache-Control:no-cache\r\n" +
+		"\r\n"
+	resp := &Response{}
+	bPool := bufiopool.New(1, 1)
+	br := bPool.AcquireReader(strings.NewReader(s))
+	byteBuffer := bytebufferpool.MakeFixedSizeByteBuffer(100)
+	bw := bufio.NewWriter(byteBuffer)
+	err := resp.WriteTo(bw)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+
+	sHijacker := &hijacker{}
+	resp.SetHijacker(sHijacker)
+
+	err = resp.ReadFrom(false, br)
+	if resp.GetSize() != 43 {
+		t.Fatal("Response read from reader failed")
+	}
+	if resp.ConnectionClose() {
+		t.Fatal("Response connection close is wrong")
+	}
+	defer bw.Flush()
+}
+
+func TestHTTPResponseError(t *testing.T) {
+	errS := "HTTP/1.1 ok\r\n" +
+		"Cache-Control:no-cache\r\n" +
+		"\r\n"
+	resp := &Response{}
+	bPool := bufiopool.New(1, 1)
+	br := bPool.AcquireReader(strings.NewReader(errS))
+	byteBuffer := bytebufferpool.MakeFixedSizeByteBuffer(100)
+	bw := bufio.NewWriter(byteBuffer)
+	err := resp.WriteTo(bw)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+
+	sHijacker := &hijacker{}
+	resp.SetHijacker(sHijacker)
+
+	err = resp.ReadFrom(false, br)
+	if !strings.Contains(err.Error(), "fail to read start line of response") {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	err = resp.WriteTo(bw)
+	if err == nil {
+		t.Fatalf("expected error: %s", err)
+	}
+	if !strings.Contains(err.Error(), "response already initialized") {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	defer bw.Flush()
+}
+
+func TestHostInfo(t *testing.T) {
```

Comment:
more tests on host info should be made after @daizong finished his PR on url rewrite